sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala

```diff
@@ -148,6 +148,18 @@ abstract class QueryStageExec extends LeafExecNode {
  */
 abstract class ExchangeQueryStageExec extends QueryStageExec {
 
+  /**
+   * AQE can trigger stage cancellation before doMaterialize function is called. This causes
```

Comment:
is it possible to avoid this? The AQE framework should know if a stage has been submitted or not.

Reply:
Firstly, thanks for the review. Yes, that is possible. Current logic iterates over `currentPhysicalPlan` for `ExchangeQueryStageExec` and try to cancel stages without checking if the stage materialization is started or not. I have just added this check as well by the last commit. After `isMaterializationStarted()` check is added to `cleanUpAndThrowException()`, `isMaterializationStarted()` check becomes redundant on `ShuffleQueryStageExec.cancel()` and `BroadcastQueryStageExec.cancel()` but i think they can still be worth if there is direct call to stage cancellation without checking` isMaterializationStarted()` for the long-term. This aims additional protection on the cancelation API but it is also up-to-us.